File: algorithms/cpp/414.cpp
#include <iostream>
#include <vector>
#include <set>
using namespace std;

class Solution {
public:
    int thirdMax(vector<int> &nums) {
        set<int> top3;
        for (auto num: nums) {
            if (top3.insert(num).second && top3.size() > 3)
                top3.erase(top3.begin());
        }
        return top3.size()==3 ? *top3.begin() : *top3.rbegin();
    }
};

int main() {
    return 0;
}

File: algorithms/python/414.py
import heapq

class Solution:
    def thirdMax(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        nums = list(set(nums))
        if len(nums) < 3:
            return max(nums)
        h = nums[:3]
        heapq.heapify(h)
        for i in range(3, len(nums)):
            heapq.heappush(h, nums[i])
            heapq.heappop(h)
        return heapq.heappop(h)
